### PR TITLE
Local files will no longer be downloaded but instead opened with a command like xdg-open for Linux, open for Mac, and start for Windows

### LINKED ISSUE #6549: FR: don't download local files, open them with their default program.
After clicking on a file that cannot be opened by qutebrowser on a local directory page (`file://` URLs), `qutebrowser` should not start a download for that local file, but instead try to open it with the default handler.

IRC conversation:
```
17:29 <emanuele6> if i'm viewing a local directory in qutebrowser
17:29 <emanuele6> and i click on a file that qutebrowser can't open, qutebrowser tries to download it
17:29 <emanuele6> e.g. mkv files
17:29 <emanuele6> is there a way to tell qutebrowser to open it in mpv?
17:32 <The-Compiler> emanuele6: nope, that behavior is coming from the underlying Chromium
17:33 <The-Compiler> emanuele6: might make sense for qutebrowser to not accept the download and open the file:/// URL with the default handler though - mind opening an issue?
```

### PR DESCRIPTION
**Pull request by jso8910**
Closes #6549

### CODE REVIEW

File: misc/requirements/requirements-pyinstaller.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # This file is automatically generated by scripts/dev/recompile_requirements.py
 
 altgraph==0.17
-pyinstaller==4.3
+pyinstaller==4.4
```
Comment (The-Compiler) [inline misc/requirements/requirements-pyinstaller.txt:4]: unrelated change

**Review by The-Compiler — CHANGES_REQUESTED** [review]

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for the quick update! I added some more comments I didn't notice before, mainly around using the [QUrl](https://doc.qt.io/qt-5/qurl.html) methods instead of treating the URL as a string and operating on that.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for the quick update! I added some more comments I didn't notice before, mainly around using the [QUrl](https://doc.qt.io/qt-5/qurl.html) methods instead of treating the URL as a string and operating on that.

edit: Whoops, got a HTTP 500 from GitHub when I tried posting this comment for the first time... Time for a [HTTP cat](https://http.cat/):

cat

File: misc/requirements/requirements-pyinstaller.txt (hunk 1 above)
Comment (The-Compiler) [reply misc/requirements/requirements-pyinstaller.txt]: Yep! Those get [updated automatically each week](https://github.com/qutebrowser/qutebrowser/pulls?q=is%3Apr+author%3Aqutebrowser-bot), so usually `recompile_requirements.py` only needs to be run manually when adding a new dependency (or removing on).

**Comment by jso8910** [comment]
I will be quite quick for the time being as I have nothing better to do (in fact that is why I picked up this pull request).

Just killing time until I go to sleep in a few hours.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Looks good overall now, but there's still a problem: When `xdg-open` doesn't know how to handle a file, it defaults to opening it in a browser. However, now qutebrowser opens `xdg-open`, which results in an endless loop of qutebrowser opening itself :sweat_smile: 

To fix that, we should check whether the download really originated from a `file:` page (which will also be a good additional security precaution).

There's [QWebEngineDownloadItem::page()](https://doc.qt.io/qt-5/qwebenginedownloaditem.html#page), so something like:

```python
origin = qt_item.page().url() if qt_item.page() else QUrl()  # at the top, where `url` is set
# ...
if url.scheme() == "file" and origin.isValid() and origin.scheme() == "file":
```

should hopefully help, though I didn't test it.

**Comment by The-Compiler** [comment]
Thanks!
